arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -1078,104 +1078,136 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
+                        _config.master.endpoint);
+        }
       }
-    }
 
-    double waitTime = TRI_microtime() - startTime;
+      double waitTime = TRI_microtime() - startTime;
 
-    if (static_cast<uint64_t>(waitTime * 1000.0 * 1000.0) >= _config.applier._initialSyncMaxWaitTime) {
-      return Result(TRI_ERROR_REPLICATION_NO_RESPONSE,
-                    std::string(
+      if (static_cast<uint64_t>(waitTime * 1000.0 * 1000.0) >= maxWaitTime) {
+        return Result(TRI_ERROR_REPLICATION_NO_RESPONSE,
+                      std::string(
                         "timed out waiting for response from master at ") +
-                        _config.master.endpoint);
+                      _config.master.endpoint);
+      }
+
+      if (isAborted()) {
+        return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
+      }
+
+      std::chrono::milliseconds sleepTime = ::sleepTimeFromWaitTime(waitTime);
+      std::this_thread::sleep_for(sleepTime);
     }
 
-    if (isAborted()) {
-      return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
+    if (replutils::hasFailed(response.get())) {
+      return replutils::buildHttpError(response.get(), url, _config.connection);
     }
 
-    std::chrono::milliseconds sleepTime = ::sleepTimeFromWaitTime(waitTime);
-    std::this_thread::sleep_for(sleepTime);
-  }
+    Result r = replutils::parseResponse(builder, response.get());
 
-  if (replutils::hasFailed(response.get())) {
-    return replutils::buildHttpError(response.get(), url, _config.connection);
-  }
+    if (r.fail()) {
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    std::string("got invalid response from master at ") +
+                    _config.master.endpoint + url + ": " + r.errorMessage());
+    }
 
-  VPackBuilder builder;
-  Result r = replutils::parseResponse(builder, response.get());
+    slice = builder.slice();
+    if (!slice.isObject()) {
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    std::string("got invalid response from master at ") +
+                    _config.master.endpoint + url + ": response is no object");
+    }
 
-  if (r.fail()) {
-    return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  std::string("got invalid response from master at ") +
-                      _config.master.endpoint + url + ": " + r.errorMessage());
-  }
+    return Result();
 
-  VPackSlice const slice = builder.slice();
-  if (!slice.isObject()) {
-    return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  std::string("got invalid response from master at ") +
-                      _config.master.endpoint + url + ": response is no object");
+  };
+
+  auto ck = keysCall(true);
+  if (!ck.ok()) {
+    return ck;
+  }
+  if (!slice.hasKey("id")) { // we only have count
+    VPackSlice const c = slice.get("count");
+    if (c.isNumber()) {
+      maxWaitTime = c.getNumber<uint64_t>() * 8 / 100000;
```

Comment:
IMHO `maxWaitTime` should be in microseconds, not in seconds.
If `count` contains a value less than 12.5K, the wait time will always be 0 here.
I think it should also have a minimum lower bound, e.g. `maxWaitTime = std::min(maxWaitTime, calculatedValue)`.